Enforce optimistic locking for user resources

rosstimothy opened the pull request:
This also ensures that SSO users created by the GitHub connector are setting the revision if the user already existed.

Contributes to https://github.com/gravitational/teleport/issues/30416
Depends on https://github.com/gravitational/teleport.e/pull/2304

fspmarshall reviewed (commented):
I haven't manually tested this change, but I believe it will cause some confusing behavior with `tctl create`.  As currently implemented mismatched revision will still fail even if `--force` is set, and `tctl`'s weak existence check will still fail even if there is the revision _is_ correct.  Imo the correct behavior should be:

- With `--force`: update always succeeds.
- without `--force`: update succeeds if user doesn't exist, _or_ revision matches (*edit*: or maybe create fails if user does not exist but revision is set and vise-versa, I suppose that'd be more internally consistent).

fspmarshall reviewed (approved):
Long-term my suggestion would be that we have empty revision always mean create and wildcard revision always mean upsert so that we can eliminate unnecessary round-trips for create/update/upsert, but I think this is a real meaningful improvement as-is.

AntonAM commented on lib/services/local/users.go:
```diff
@@ -238,17 +244,18 @@ func (s *IdentityService) UpsertUser(user types.User) error {
 	}
 	if auth := user.GetLocalAuth(); auth != nil {
-		if err = s.upsertLocalAuthSecrets(user.GetName(), *auth); err != nil {
+		if err = s.upsertLocalAuthSecrets(context.TODO(), user.GetName(), *auth); err != nil {
```
We already use `context.TODO()` in this function, might extract it to `ctx := context.TODO()` like in `CreateUser`

rosstimothy committed:
Enforce optimistic locking for user resources

This also ensures that SSO users created by the GitHub, OIDC, and
SAML connectors are setting the revision if the user already existed.

Contributes to #30416
Depends on gravitational/teleport.e#2304
